### PR TITLE
Support gradient with coordinate

### LINKED ISSUE #3945: Gradient with coordinate
Current version of dask.Array.gradient [does not support non-uniform coordinate](https://github.com/dask/dask/blob/5fa703eaaa6679e452c1dc7cdc8c9ac039994de5/dask/array/routines.py#L437-L438).
I implemented this in pydata/xarray#2398, but I think it would be better to be integrated in the upstream.
Is it in scope of dask development?

cc. @jakirkham

### PR DESCRIPTION
**Pull request by fujiisoup**
- [x] Tests added / passed
- [x] Passes `flake8 dask`

Now da.Array.gradient supports coordinate.
Closes #3945

### CODE REVIEW

**Comment by mrocklin** [comment]
cc @jakirkham , you might find this interesting

**Comment by TAdeJong** [comment]
How does this compare performance wise if *no* coordinate arg is supplied? Both in terms of speed and memory usage it might be useful to create an internal 'None' case to circumvent the creation of the `array_loc_start` and `array_loc_end` arrays.

**Comment by jakirkham** [comment]
Thanks for putting this in, @fujiisoup 😄

Agree with @TAdeJong that we may want to handle some spacing cases differently (e.g. uniform). Though a benchmark with NumPy should be sufficient for determining whether this is actually needed.

**Comment by fujiisoup** [comment]
Thanks, @TAdeJong, @jakirkham.

> How does this compare performance wise if no coordinate arg is supplied?

Thanks for the good suggestion. Updated. Now `array_loc_start` and `array_loc_end` are not created for the scalar coordinate case.

For the benchmarking, I think my modification does not degrade a lot, but not yet confirmed.
Is there a good way to make a benchmarking with dask?

**Comment by jakirkham** [comment]
Thanks @fujiisoup.

Will give this a closer look tomorrow, but this seems reasonable from a quick glance.

Don't think we need to worry about benchmarking if uniform grids are special cased. The concern here was only if uniform grids were handled the same as non-uniform grids, which is no long true.

File: dask/array/routines.py [hunk 1]
```diff
@@ -465,26 +469,54 @@ def gradient(f, *varargs, **kwargs):
...
 
-    r = [
-        f.map_overlap(
+    results = []
+    for i, ax in enumerate(axis):
+        for c in f.chunks[ax]:
+            if c < kwargs["edge_order"] + 1:
+                raise ValueError(
+                    'Chunk size must be larger than edge_order + 1. '
+                    'Given {}. Rechunk to proceed.'.format(c))
```
Comment (jakirkham) [inline dask/array/routines.py]: Maybe we should mention the axis this was taken from?

File: dask/array/routines.py [hunk 2]
```diff
@@ -465,26 +469,54 @@ def gradient(f, *varargs, **kwargs):
...
+
+        if np.isscalar(varargs[i]):
+            array_locs = None
+        else:
+            # coordinate position for each block taking overlap into account
+            array_loc_stop = np.cumsum(np.array(f.chunks[ax])) + 1
+            array_loc_start = array_loc_stop - np.array(f.chunks[ax]) - 2
```
Comment (jakirkham) [inline dask/array/routines.py]: Can probably refactor out `np.array(f.chunks[ax])` as both of these use it.

**Comment by jakirkham** [comment]
Made some minor comments. Still looks pretty good.

One question came up while I was thinking about this. Should we try to handle coordinates that are Dask Arrays? FWIW don't think this would be too difficult. Just a question of whether this would be useful.

**Comment by fujiisoup** [comment]
Thanks, @jakirkham
Updated.

> Should we try to handle coordinates that are Dask Arrays?

Although xarray always keeps coordinate as np.array and may not use it, dask.array coordinate support sounds reasonable.

I thought my PR is already working with dask.array coordinate, but the additional test with dask coordinate is failing with claiming that
```python
>       r_d_a = da.gradient(d_a, *varargs, axis=axis, edge_order=edge_order)

tests/test_routines.py:468: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
routines.py:517: in gradient
    grad_kwargs=kwargs,
core.py:1884: in map_overlap
    return map_overlap(self, func, depth, boundary, trim, **kwargs)
overlap.py:463: in map_overlap
    g2 = g.map_blocks(func, **kwargs)
core.py:1828: in map_blocks
    return map_blocks(func, self, *args, **kwargs)
core.py:739: in map_blocks
    dsk[k] = dsk[k][:-1] + (assoc(dsk[k][-1], 'block_id', k[1:]),)
cytoolz/dicttoolz.pyx:354: in cytoolz.dicttoolz.assoc
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   AttributeError: 'list' object has no attribute 'keys'

cytoolz/dicttoolz.pyx:368: AttributeError
```

map_overlap does something for dask.Array coordinate?

**Comment by jakirkham** [comment]
Thanks for sharing.

Basically Dask Arrays need to be arguments instead of keyword arguments. They may need some singleton dimensions as a hint for broadcasting too. Think that should be sufficient to fix it.

That said, wouldn’t worry about it too much as it doesn’t seem to be a useful case for you. We can always save it in an issue and see if someone else comes along and picks it up.

**Comment by jakirkham** [comment]
Thanks @fujiisoup 😄

**Pull request merged**